stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/adapter/viewholder/decorator/BaseDecorator.kt

```diff
@@ -0,0 +1,58 @@
+package io.getstream.chat.android.ui.messages.adapter.viewholder.decorator
+
+import com.getstream.sdk.chat.adapter.MessageListItem
+import io.getstream.chat.android.core.internal.exhaustive
+import io.getstream.chat.android.ui.messages.adapter.BaseMessageItemViewHolder
+import io.getstream.chat.android.ui.messages.adapter.viewholder.MessageDeletedViewHolder
+import io.getstream.chat.android.ui.messages.adapter.viewholder.MessagePlainTextViewHolder
+import io.getstream.chat.android.ui.messages.adapter.viewholder.OnlyFileAttachmentsViewHolder
+import io.getstream.chat.android.ui.messages.adapter.viewholder.OnlyMediaAttachmentsViewHolder
+import io.getstream.chat.android.ui.messages.adapter.viewholder.PlainTextWithFileAttachmentsViewHolder
+import io.getstream.chat.android.ui.messages.adapter.viewholder.PlainTextWithMediaAttachmentsViewHolder
+
+internal abstract class BaseDecorator : Decorator {
+    final override fun <T : MessageListItem> decorate(viewHolder: BaseMessageItemViewHolder<T>, data: T) {
+        if (data !is MessageListItem.MessageItem) {
+            return
+        }
+        when (viewHolder) {
+            is MessageDeletedViewHolder -> decorateDeletedMessage(viewHolder, data)
+            is MessagePlainTextViewHolder -> decoratePlainTextMessage(viewHolder, data)
+            is OnlyMediaAttachmentsViewHolder -> decorateOnlyMediaAttachmentsMessage(viewHolder, data)
+            is PlainTextWithMediaAttachmentsViewHolder -> decoratePlainTextWithMediaAttachmentsMessage(viewHolder, data)
+            is OnlyFileAttachmentsViewHolder -> decorateOnlyFileAttachmentsMessage(viewHolder, data)
+            is PlainTextWithFileAttachmentsViewHolder -> decoratePlainTextWithFileAttachmentsMessage(viewHolder, data)
+            else -> Unit
+        }.exhaustive
+    }
+
+    protected abstract fun decoratePlainTextWithFileAttachmentsMessage(
```

Comment:
Why did you make all these func abstract?

Reply:
Realized that having default implementations makes it easy to forget to add one of the overrides in a decorator (because I've done it the first time around). Feels better to be forced to make the decision that you want to override with an empty method.

Reply:
I think it's kind of trade-off. None of these is the best (abstract and with empty impl). Why did then you add abstract for  decorate message deleted? Seems deleted message never requires to get decorated

Reply:
I think being more explicit is better in this case, makes it harder to forget to decorate something. The kinds of ViewHolders we have are not completely obvious, e.g. someone might not remember that we have a separate one for `PlainTextWithFileAttachments` or something. Same goes for adding a new ViewHolder type, you might forget to update one of the decorators otherwise.

For the deleted ViewHolder: that might be the case, but making one method non-abstract would make it even trickier to work with the class. Based on all the others, you would assume that you're forced to implement the method for all ViewHolders, and then this one would be a surprising exception to that pattern. Consistency makes for safer code and fewer bugs 😄 .

Reply:
I may agree about others, but such view holders as date divider, message deleted I'm not sure it worths to make such methods abstract

Reply:
Alright, when we have more than one of those kinds of ViewHolders, we can add default implementations for them, then it's more obvious that some are required and some are not. (Though personally, I'd prefer to keep things one way, since it's not that verbose to do.)